### ISSUE TITLE
Recommendation for exposing multiple TypeScript modules from single NPM package

### ISSUE THREAD

**Issue by joewood** [Microsoft/TypeScript#8305]
The module resolution logic in the handbook doesn't mention how nested modules are supported from a single npm package. There is an old issue discussing this that has been closed:  https://github.com/Microsoft/TypeScript/issues/5804

Neither the solution here or simply trying to refer to modules beneath the root dir referred to by the **typings** or **main** path seems to work. Is there a documented best practice? The documentation says that ambient modules should not be used in the **typings** field in **package.json**.

Just to be clear, I want to write this in a consuming app:
```typescript
import {stuff} from "my-npm-lib/submodule"
```

Where the package.json file in **my-npm-lib** has:
```json
    "main": "lib/index.js",
    "typings": "lib/index.d.ts",
```

And the file structure in the **my-npm-lib**:
```
my-npm-lib\
    index.js
    index.d.ts
    submodule.js
    submodule.d.ts
```

Currently this complains with a module not found error. But this works, but shouldn't:
```typescript
import {stuff} from "my-npm-lib/lib/submodule"
```

 This is all with the module resolution setting set to **node**.

**Comment by mhegazy** [comment]
We need to add documentation, so leaving this issue open to track that, and marking it as such.

For the question, the recommendation is to generate the .d.ts with 1:1 mapping to your .js files, and place them next to the .js file in the published package, as you mentioned. the `"typings"` field specify the main entry point, matching `"main"` for the .js file.

I have a sample posted at https://github.com/mhegazy/npm-dependency-test, take a look and let me know if you have more questions.

**Comment by joewood** [comment]
Thanks @mhegazy 
As I thought. The only real way of achieving submodules is to pollute the root directory. I think that's a limitation of JavaScript/npm too though, so not a massive deal. 

One option to avoid the pollution is to use a `postinstall` script to only pollute the root dir after **npm install**. This works fine, but it doesn't work too well with **npm link**. Another option could be to just move the build files from a distribution dir to the root in postinstall. Anyway, right now I've just resorted to adding a `files.exclude` filter in **vscode** to hide the mess.

As you suggested, it would be great to update the handbook on best practices for distributing typescript built libraries, especially with respect to **typings** and dependent types.

**Comment by jcansdale** [comment]
@mhegazy I'm confused about why there is the need for a separate "typings" entry when the .d.ts is invariably next to the one referenced by "main". Why not simply look next to the "main" .js or at least look there as well?

**Comment by mhegazy** [comment]
Not sure i see the issue rely. if you are adding `"main"` property in your package.json, should not be too much to ask to add `"typings"` as well.

**Comment by jcansdale** [comment]
@mhegazy It's more that is seems inconsistent. From my understanding, if you: import foo = require("foo.js"), TypeScript tools will automatically look for type definitions in "foo.d.ts". If however you: require("bar") and there is a "main": "foo.js" in bar/package.json, TypeScript tools won't automatically look for type definitions in "bar/foo.d.ts".

This is a shame because otherwise you could convert a JavaScript package into TypeScript by simply renaming the .js files to .ts and compiling with the definitions option. Unfortunately you need to know about the "typings" option otherwise your consumers won't see your type definitions. It just seems like an unnecessary gotcha. Why not support "main" relative as well as "typings"?

**Comment by mhegazy** [comment]
it does work for index.js -> index.d.ts. We could add support for `main`. but the idea was if you chose to change the defaults for your package by setting a `main`, then you should also consider what you want to put in `typings`, they may or may not be in the same place.

**Comment by jcansdale** [comment]
How about making it so that if `typings `exists look for it there, otherwise look for it relative to `main`? Maybe there is one but I can't think of a good reason why you wouldn't want to look next to `main` if a `typings` entry doesn't exist. I'm thinking more for package authors who've written their package using TypeScript, rather than authors who want to add typings to their JavaScript package.

**Comment by mhegazy** [comment]
feel free to file a separate issue for this proposal.

**Comment by bogdanmanate** [comment]
This will definitely be a major improvement for Typescript development. I'm currently developing a library in TS and this library should be included in an Angular 2 application. I've also looked into angular repository to see how they are exporting the modules, but it seems that they bundle the modules in a single file.

**Comment by mhegazy** [comment]
here is the documentation for writing declaration files. please let us know if there are any missing scenarios.

**Comment by bogdanmanate** [comment]
Hi @mhegazy and thank you for your answer. Where can I find the documentation ?

**Comment by mhegazy** [comment]
Apologies. forgot to add the link. here you go: https://github.com/Microsoft/TypeScript-Handbook/blob/master/pages/declaration%20files/Introduction.md

also fixed the original comment.

**Comment by pksorensen** [comment]
My usecase is the following, my project builds and output the generated d.ts files to

`/dist/typings`

and I then in my package.json have specificed that `types:"dist/typings/index.d.ts"`.

From consumer project the following work and it consumes the `/dist/typings/index.d.ts` file

`import * as doh from "libproject"`

but when doing 

`import * as foo from "libproject/test"`

it cannot find the `/dist/typings/foo/index.d.ts` file that was also generated as part of the `libproject`.

Copying stuff to `node_modules/libproject/dist/typings` to `node_modules/libproject/` as part of install is not really a clean and nice solution.

**Comment by bogdanmanate** [comment]
@pksorensen I'm using the same workflow, but I'm also finding it not so elegant and clean. It will be nice to have a method/process that allows you can export a single file (or archive) and than you're good to go. The developers that have already worked with JAVA may relate this to a library export as a .jar file.

**Comment by pksorensen** [comment]
I found a better approach. In my consumer project wiht typescript 2.0 I could set it up to resolve submodules.

See the compile time section of readme at https://github.com/s-innovations/kolayout , where setting path compile option of tsc to resolve "mylib/*" to the dist folder of d.ts files of my mylib.

**Comment by Meligy** [comment]
I have the same flow as described in https://github.com/Microsoft/TypeScript/issues/8305#issuecomment-254017287

That is, using a subfolder for generated files (.js and .d.ts).

When you do that, and you have:

```
- index.ts
- sub-module
  - index.ts
- built
  - index.js
  - index.d.ts
  - sub-module
    - index.js
    - index.d.ts
```

Setting the `types` and / or `typings` property in `package.json` to `built/index.js` makes this work:

```
import * as mainLib from 'main-lib';
```

However, this still fails:

```
import * as subModule from 'main-lib/sub-module';
```

(Of course this works `import * as subModule from 'main-lib/built/sub-module';` but is undesirable)

One suggestion above was compiling to the same source directory.
Is there a way to make it work if I compile to a sub-directory?

**Comment by mhegazy** [comment]
> Is there a way to make it work if I compile to a sub-directory?

no there is not currently.. 

We could in theory do multiple lookups if `main-lib/sub-module` failed to load. first go up to `main-lib` and resolve `sub-module` as a module of it, i.e. resolve `main-lib` first, then use that as a base for finding `sub-module`. that would be a new feature though, and we would diverge from eh node resolution logic. i would rather we did not do that, so many assumptions are already backed in based on the node resolution logic.

**Comment by Rustygames** [comment]
I'd really like to see this reopened and reconsidered. Module resolution always working from the root folder doesn't make sense in any use case. The preference is always to resolve from the lib or dist directory.

Even if this has to be added as a feature which departs from node resolution logic (which I don't really see how it does), then I still think it's essential

**Comment by alexander-alvarez** [comment]
Chiming in on this because I lost an hour or so figuring this out.

Pushing it onto the consumer to define the submodules in their tsconfig is not a good thing when the responsibility to define those lies with the library owner.

That being said I library I know used to have this problem made it do away with this PR https://github.com/chriskrycho/true-myth/pull/15

They used to have the disclaimer for sub modules
https://github.com/chriskrycho/true-myth/pull/15/files#diff-04c6e90faac2675aa89e2176d2eec7d8L158
and they're only using Typescript to build -- https://github.com/chriskrycho/true-myth/pull/15/files#diff-b9cfc7f2cdf78a7f4b91a753d10865a2R33

I think it has to do with the ["module": "ESNext"](https://github.com/chriskrycho/true-myth/pull/15/files#diff-9acfb03ff602669d71700e9cbdf1dc5cR4) but I'm still figuring it out.

maybe @chriskrycho can chime in and help us all out

**Comment by chriskrycho** [comment]
I wrote up the problem and solution [here](https://www.chriskrycho.com/2018/how-to-bundle-typescript-type-definitions.html). I definitely think that it would be extremely useful to be able to generate single-file module-supporting definitions, or to do type lookups for modules somewhere besides the root.

**Comment by Rustygames** [comment]
@chriskrycho nice write up thank you.

I'd still like to see it support types from a dist/lib/build folder rather than root so that you don't have to rely on prepublish hook 'hacks'.
Then you can, for example, point to the library on your file system for fast editing, rather than having to do a publish every time

**Comment by marcj** [comment]
This could probably be solved via a `mainDir` configuration that is in discussion https://github.com/nodejs/node/issues/14970

**Comment by mrmckeb** [comment]
I see this issue was closed off, and there doesn't seem to be any consensus on how to move forward.

Imagine I have a component library, and I want TypeScript to _suggest_ importing `Button` from `some-library/lib/components/super-nested/Button`.

The user could import `Button` by typing the full path, but only imports from root are suggested by TypeScript - typing `Butto...` would not suggest anything, unless Button was also exported at root.

If I'm missing something obvious, please let me know!

**Comment by chriskrycho** [comment]
I believe (and am going to be exploring further this way) that [API Extractor](https://api-extractor.com/pages/overview/demo_rollup/) will let us do what we need this way.

**Comment by joewood** [comment]
I'm still trying to understand the usecase here. Why doesn't importing from root work in this case? Where a single index.d.ts file exports from various other declaration files from the generated types? Now that webpack can tree shake with ESM modules we switched away from deep path imports.

**Comment by chriskrycho** [comment]
Just because you *can* do it that way doesn’t mean everyone *wants* to design things that way. I take module design to be an important part of API design, and therefore structure my imports accordingly – usually a single default export and a number of supporting named exports, like a data at structure and helper functions or secondary types that go with it.

An `index.d.ts` file *can* support this way of structuring an API and module graph, using `declare module`, but it has to be either handwritten or generated by something like API extractor to do so: there is no emit mode in TS’ own type generation that supports it.

**Comment by joewood** [comment]
I guess I'm missing what the driver is to use deep paths over named exports. For me, named exports from a single root are more discoverable using the type info. What's the motivating factor?

**Comment by chriskrycho** [comment]
If you have a small enough library, maybe that makes sense. For a relatively small example, see the import structure of my [True Myth](https://github.com/true-myth/true-myth) library, which has several discrete modules with their own helper functions, *some of whose names are the same*! For example:

```ts
import Maybe, { map } from 'true-myth/maybe';
```
```ts
import Result, { map } from 'true-myth/result';
```

For a much larger example, see the import design for [Ember](https://api.emberjs.com/ember/release).

In short, if your library is larger than a handful of imports, or if there are names that are the same across modules (which is often perfectly reasonable, as in the example I gave above!), throwing them all into the root just isn't an option. Again: *modules themselves* are often part of the API design. It's not *named exports **or** deep paths*, it's often *named exports **and** deep paths*.

**Comment by mrmckeb** [comment]
> If you have a small enough library, maybe that makes sense
Correct, we have a library that is shared (internally) and it's quite big. We could export everything at root, but it would certainly be cleaner to have "submodules".

I know we could also just separate these out into different packages as an alternative, but that complicates local development in our workflow.

**Comment by sweeetland** [comment]
Almost four years later and still no practical solution to this? 🤔

**Comment by tomsh99** [comment]
Any solution for this frustrating problem?

**Comment by vankessel** [comment]
Ditto :point_up: Would be nice to be able to do this

**Comment by jaens** [comment]
This used to be a fundamental restriction of Node.js, but recently there's added support for the `exports` field in `package.json` which does allow this.

TypeScript support is tracked in #33079.

**Comment by panva** [comment]
> Any solution for this frustrating problem?

https://github.com/microsoft/TypeScript/issues/33079#issuecomment-702617758

**Comment by slavafomin** [comment]
I've also spent a lot of time today trying to figure out why my declaration files are not loaded. I was compiling my declarations to the `dist/types/*.d.ts` directory. I was expecting TypeScript to be able to resolve type modules relative to the `dist/types/index.d.ts` file specified in package.json: `"types": "dist/types/index.d.ts"`. However, it wasn't the case. I've tried to use the following option: `"types": "dist/types/"` to specify the root directory, but it wasn't working either.

I guess it would be very reasonable to expect from TypeScript to look for declarations relative to the directiry/file specified in `types` property of `package.json` manifest.

I can't quite comprehend why this issue is closed without a solution…

**Comment by Toliak** [comment]
@slavafomin, the same... I have reproduced this bug (I think it is bug...) with `rxjs@7.0.0-beta.12`.
There is [types in package.json](https://github.com/ReactiveX/rxjs/blob/a4b4e998d6bd86c8099fb626b9f5d26e8a8137fc/package.json#L8)
However, auto import suggests nothing for example for `pipe` (and this variable is exported from `index.d.ts`)

I will upload screenshots later...

**Comment by tstewart-klaudhaus** [comment]
Thanks to @chriskrycho for the very clear description of the situation and workaround in his article linked from https://github.com/microsoft/TypeScript/issues/8305#issuecomment-391929338

The fact that the TS compiler cannot resolve submodule definitions that were generated and positioned within the package structure by the TS compiler itself is pretty remarkable. As a case of the TS compiler not being able to consume its own output, it could be viewed as an internal inconsistency.

My use case is providing a submodule with interfaces intended to be augmented (merged) via ambient declaration. This is itself a workaround for TS lack of higher-kinded types. I get why the HKT thing is difficult, but this submodule resolution issue seems on the surface like it should be much more tractable.

**Comment by softmarshmallow** [comment]
I'm having same problem while using one package for both npm publication and for using as submodule package as my other typescript project.

specifying main should point index.ts file and tsc or yarn/npm should somehow automatically link to dist/index.js

or maybe this should be another thread fro npm & package.json?

**Comment by Stvad** [comment]
https://github.com/microsoft/TypeScript/issues/33079 seems to be fixed, but I still can't find a great working example of a library with submodules hierarchy set up. Anyone has good references?

**Comment by ShaMan123** [comment]
I have spent hours trying to configure a repo to work with submodule types. The repo has a build tool, rollup, and a folder structure that doesn't reflect the submodule importing scheme.

https://github.com/microsoft/TypeScript/issues/8305#issuecomment-793715471 has pointed out rxjs so I looked at the config they've done since they manage to expose typed submodules with a build tool and w/o mirroring the submodule importing scheme as suggested in the [official guide](https://www.typescriptlang.org/docs/handbook/declaration-files/templates/module-d-ts.html#library-file-layout).

Related: #17945 

### Findings
It seems that though TS doesn't respect the `exports.*.types` field in `package.json` it does respect using `typesVersions` to specify submodule resolution strangely enough (e.g. https://github.com/fabricjs/fabric.js/pull/8632/commits/b326dd6033aac1a285daed1ae090736aaff8a366)
I am not sure if this is enough on its own or if you need to define `paths` in `tsconfig.json` so I did that as well.
I am sure that defining `paths` in `tsconfig.json` w/o the `package.json` config doesn't work (though I read it should).
Also it seems relative paths are not recognized so use abs syntax that is relative to root.

Hope this helps someone.
TS if it works with `typesVersions` sounds like it can work with the `exports.*.types` field in `package.json`.
